spec/ruby/core/thread/join_spec.rb

```diff
@@ -19,8 +19,13 @@
     t.join(0).should equal(t)
     t.join(0.0).should equal(t)
     t.join(nil).should equal(t)
+  end
+
+  it "raises TypeError if the argument is not a valid timeout" do
+    t = Thread.new {sleep}
     -> { t.join(:foo) }.should raise_error TypeError
     -> { t.join("bar") }.should raise_error TypeError
+    t.kill
```

Comment:
This can leak a Thread outside this spec example, it needs `t.join` after `t.kill`.

Also your spec change means things that worked fine before now fail randomly, including the previous version of this spec but maybe other code too.

Reply:
I fixed both (c19bcf38ebe4f483bdb2d76f2115eb70f056f8c6 and 82998918efa3a637c80e135198b573cfcb31acd9).
Please take the time to keep existing behavior and not introduce extra non-determinism next time.

Reply:
The behaviour of `t.join` when the thread is dead is not specified anywhere. Ruby is a dynamically typed language and also lazy in many places. So I don't think that a type coercion error is required if don't need the results.

Reply:
> The behaviour of `t.join` when the thread is dead is not specified anywhere

It's tested by this spec. There is no official specification document of Ruby besides the ISO one which is outdated and covers a rather small subset.

I guess at least we can agree flakey/non-deterministic tests and behavior is bad? If so, the previous behavior is more deterministic, coercion always happens, whether or not the target thread concurrently joined.

> Whether we validate types or not is something we should decide on based on merit.

What's the merit to change behavior incompatibly like this?

> I'm in favour of only doing it if required because it's more efficient.

Really? Spawning a thread likely takes 1000x longer than that coercion. And it's not the real reason for this change, it's you consider keeping compatibility in this case "busy work", and I think that's not a good reason to change behavior.

> In this case it's just busy work with no value to anyone

Clearly it has value to me, because I spent time to fix that and restore compatibility/determinism.

> I'm sure there is precedent for this in other parts of Ruby

Any example?

> I don't feel that one way or another is standardised.

It's something worth testing IMHO.
And I guess the expectation of most users is argument coercion/validation happens before anything else.
It's also seems a good pattern to me in general, that way there is no late coercion that might mess up some cleanup or so.

> The implementation you gave essentially repeats the same operation twice,

Actually no, there is first `Ruby->nil | Fixnum | Float` and then later conversion to `hrtime` (if needed, that has no side effects). It's not very nice in the source, I'll give you that.

> thread.join(Float::INFINITY)

Waits for an infinite duration on TruffleRuby, same on CRuby & JRuby (seems more intuitive to me than an exception).

> Can we make this consistent?

I agree that would be nice, could you make a feature request on https://bugs.ruby-lang.org/ ?
FWIW TruffleRuby has some logic to deduplicate such coercion in https://github.com/oracle/truffleruby/blob/4883714a6ec7213b99d345b2bafe13433fca212c/src/main/java/org/truffleruby/core/cast/DurationToMillisecondsNode.java

> Finally, this misses the core issue, and your fix breaks the assumption, that timeout arguments would be passed unchanged through to the scheduler. However, this was already broken before, so it's something I'll write specs for and hopefully we can start to make things more consistent, reliable and precise.

I was not aware that timeout arguments should be passed unchanged through to the scheduler. I don't think it's ideal, because then the scheduler would need to handle coercion on its own, and duplicate (in logic, not perf) coercion behavior which is already encoded in the C code.

`Integer | Float | nil` (and prefer `nil` to infinite Float to simplify checks?) is probably the most convenient for the scheduler, no? Or do you envision passing another type? If so, which type and would that type work with blocking operations or not (then disabling a scheduler would break code which seems suboptimal)?

Reply:
Another reason to convert eagerly:
Without a GIL/GVL and with parallel threads, one cannot know if the target Thread is terminated, it might just terminate while trying to join.
So coercion needs to happen anyway to join: https://github.com/oracle/truffleruby/blob/aa495587766ec1f21ee8c98a5d04ea3739836a1b/src/main/java/org/truffleruby/core/thread/ThreadNodes.java#L399
One could do an early check but then you get 3 cases (coerced & waited, coerced but concurrently terminated so not waited, not coerced because already terminated) and the semantics become really complex for the user.